### ISSUE TITLE
Live editor mixup between ts and tsx

### ISSUE THREAD

**Issue by mjambon** [returntocorp/semgrep#2191]
Originally reported by @msorens.

**Describe the bug**

In the live editor, a legal typescript file fails to parse if it contains a type assertion using the angle brackets syntax. However, the semgrep command line works fine. The error message is
```
Could not parse target.tsx as ts
```
Based on the error message above, it looks like the snippet was incorrectly put in a file with the `.tsx` extension instead of `.ts`. The extension is then used by semgrep-core to pick the tsx parser rather than the ts parser. The tsx parser fails on that input due to the angle bracket type assertion (`<the_type> value`) which is invalid tsx. It's actually the only typescript construct that is not also supported by tsx. This is annoying but we have to deal with it somehow. Because of this incompatibility, typescript/tsx offers an alternate syntax for type assertion: `value as the_type` is equivalent to `<the_type> value`.

**To Reproduce**

See https://semgrep.dev/s/B2Ob/

**Expected behavior**

No parse error.

**Screenshots**

![image](https://user-images.githubusercontent.com/343265/101204923-d75dc700-3621-11eb-8a7e-5c00b28ecf85.png)

**What is the priority of the bug to you?**

P2: works in semgrep, fails in live editor

**Environment**

Live editor only.

**Comment by underyx** [comment]
Oh wow, that's nasty, I was so happy about the hack of making all TypeScript snippets .tsx to support both! But then we have to change this somehow.

I just don't know how. Will we need to create a new, separate menu option to select TSX as snippet language explicitly?

**Comment by mjambon** [comment]
@underyx yes, the easy, clean way would be to provide tsx as a separate language. There may be other ways, like perhaps trying both languages speculatively and picking the one that works.

**Comment by underyx** [comment]
@msorens I considered this low priority because I never used this kind of type assertion. Could you tell me if this is something more than 1% of users would use? If it's as rare as I expect (and I might be wrong) then it might be worth it to keep the live editor UI simpler for our users and leaving only a TypeScript option in the menu.

I don't think we can try speculatively because the same logic is used for creating semgrep-rules PRs, where we don't execute semgrep.

@dlukeomalley can I ask for your advice here as well? Summary for you: when using one specific (I think very rare, but this is to be confirmed) language feature of TS, semgrep.dev fails to run rules. The only way to fix this seems to be splitting the language select menu from TS, JS to TS, TSX, JS, JSX. Do we keep the more broken but simpler UI, or do we make the UI more complex for better reliability?

**Comment by msorens** [comment]
Wow, great question, @underyx ! In Chef's code base it is very common--I count over 600 instances in our set of TS files. And those are put in not just by me 😉 , but at least a dozen people.
While, like you, I do not know if my experience can validly be extrapolated to make broader claims for other companies, my gut feeling is the opposite of yours, that anyone writing serious TypeScript would be commonly using it. I tried to find any statistics on frequency of TypeScript features being used, but no luck so far... 🤷 So not clear how to advise you here.

**Comment by mschwager** [comment]
@underyx @nbrahms do you know the status of this issue? It looks like it was moved into "To do... later" then back to "To do."

**Comment by chmccreery** [comment]
@dlukeomalley we are talking about priority of this issue- can you weigh in here?

**Comment by DrewDennison** [comment]
I looked into this and didn't see that we have different modes between ts and tsx so since tsx should be a superset, I think it's correct that all ts files are turned into tsx in the editor, if that's not correct, we should make a new language in the dropdown

**Comment by dlukeomalley** [comment]
@DrewDennison Following up, does that mean you'd like to close the ticket? 

Any thoughts from anyone else on this thread?
